File: fmea_backend/schemas/pms_plan.py
"""PMS Plan Generator API schemas (FMEA + MAUDE-like signals + AI)."""
from __future__ import annotations

from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, ConfigDict, Field


class PmsPlanGenerateRequest(BaseModel):
    project_id: str = Field(..., min_length=1)
    device_name: str = Field(..., min_length=1)
    intended_use: str = Field(..., min_length=1)


class MaudeSignalPublic(BaseModel):
    model_config = ConfigDict(extra="allow")

    failure_mode: str
    event_count: int
    trend: str
    severity: str
    source: Optional[str] = None
    notes: Optional[str] = None
    recommended_monitoring_focus: Optional[str] = None


class PmsPlanSections(BaseModel):
    """Structured PMS plan body (ISO 14971 / FDA-oriented narrative sections)."""

    device_overview: str
    pms_objectives: str
    data_sources: str
    maude_analysis: str
    risk_mapping: str
    signal_detection: str
    pms_activities: str
    capa_integration: str
    benefit_risk: str
    reporting: str


class PmsPlanGenerateResponse(PmsPlanSections):
    """Response from POST /pms/generate (matches persisted record + audit metadata)."""

    generation_id: str
    project_id: str
    created_at: datetime
    maude_signals: List[MaudeSignalPublic]
    fmea_row_count: int
    model: Optional[str] = None
    ai_generated: bool = False
    summary: str = ""
    status: str = "draft"
    version: int = 1
    warning: Optional[str] = None


class PmsPlanHistoryItem(BaseModel):
    """One saved or legacy-audited PMS plan."""

    model_config = ConfigDict(from_attributes=True, extra="ignore")

    id: str
    project_id: str
    device_name: Optional[str] = None
    intended_use: Optional[str] = None
    created_at: datetime
    input_summary: Optional[str] = None
    summary: Optional[str] = None
    status: Optional[str] = "draft"
    version: Optional[int] = None
    plan: PmsPlanSections
    maude_signals: List[MaudeSignalPublic] = Field(default_factory=list)
    fmea_row_count: Optional[int] = None
    model: Optional[str] = None
    warning: Optional[str] = None
    ai_generated: Optional[bool] = None


class PmsPlanHistoryListResponse(BaseModel):
    project_id: str
    items: List[PmsPlanHistoryItem]

File: frontend/src/api/pmsPlan.ts
/**
 * PMS Plan Generator API (FMEA + MAUDE-like signals + AI).
 * Backend: /pms/* (Pro-gated).
 */
import api, { API_BASE_URL } from '../axios';

export interface PmsMaudeSignal {
  failure_mode: string;
  event_count: number;
  trend: string;
  severity: string;
  source?: string | null;
  notes?: string | null;
  recommended_monitoring_focus?: string | null;
}

export interface PmsPlanSections {
  device_overview: string;
  pms_objectives: string;
  data_sources: string;
  maude_analysis: string;
  risk_mapping: string;
  signal_detection: string;
  pms_activities: string;
  capa_integration: string;
  benefit_risk: string;
  reporting: string;
}

export interface PmsPlanGenerateResponse extends PmsPlanSections {
  generation_id: string;
  project_id: string;
  created_at: string;
  maude_signals: PmsMaudeSignal[];
  fmea_row_count: number;
  model?: string | null;
  ai_generated: boolean;
  summary: string;
  status: string;
  version: number;
  warning?: string | null;
}

export interface PmsPlanHistoryItem {
  id: string;
  project_id: string;
  device_name?: string | null;
  intended_use?: string | null;
  created_at: string;
  input_summary?: string | null;
  summary?: string | null;
  status?: string | null;
  version?: number | null;
  plan: PmsPlanSections;
  maude_signals: PmsMaudeSignal[];
  fmea_row_count?: number | null;
  model?: string | null;
  warning?: string | null;
  ai_generated?: boolean | null;
}

export interface PmsPlanHistoryListResponse {
  project_id: string;
  items: PmsPlanHistoryItem[];
}

export async function generatePmsPlan(body: {
  project_id: string;
  device_name: string;
  intended_use: string;
}): Promise<PmsPlanGenerateResponse> {
  const { data } = await api.post<PmsPlanGenerateResponse>('/pms/generate', body);
  return data;
}

export async function listPmsPlans(projectId: string): Promise<PmsPlanHistoryListResponse> {
  const { data } = await api.get<PmsPlanHistoryListResponse>(`/pms/${encodeURIComponent(projectId)}`);
  return data;
}

export async function getPmsPlan(generationId: string): Promise<PmsPlanHistoryItem> {
  const { data } = await api.get<PmsPlanHistoryItem>(`/pms/plan/${encodeURIComponent(generationId)}`);
  return data;
}

/** Opens printable HTML in a new tab (auth via axios blob). */
export async function openPmsPlanPrintView(generationId: string): Promise<void> {
  const res = await api.get(`/pms/plan/${encodeURIComponent(generationId)}/export/html`, {
    responseType: 'blob',
  });
  const blob = new Blob([res.data], { type: 'text/html;charset=utf-8' });
  const url = URL.createObjectURL(blob);
  const w = window.open(url, '_blank', 'noopener,noreferrer');
  if (!w) {
    URL.revokeObjectURL(url);
    throw new Error('Popup blocked — allow popups or use download.');
  }
  w.addEventListener('beforeunload', () => URL.revokeObjectURL(url));
}

/** Raw URL for debugging (no auth). Prefer `openPmsPlanPrintView`. */
export function pmsPlanExportHtmlPath(generationId: string): string {
  return `${API_BASE_URL}/pms/plan/${encodeURIComponent(generationId)}/export/html`;
}
